lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance/Size/Coin.hs

```diff
@@ -0,0 +1,260 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+
+-- |
+-- Copyright: © 2024 Cardano Foundation
+-- License: Apache-2.0
+--
+-- The main purpose of this module is containing 'distributeSurplus', which is
+-- needed by 'balanceTransaction'.
+module Internal.Cardano.Write.Tx.Balance.Size.Coin
+    ( sizeOfCoin
+    , costOfIncreasingCoin
+    , maximumCostOfIncreasingCoin
+
+    -- * distributeSurplus
+    , distributeSurplus
+    , TxFeeAndChange (..)
+    , ErrMoreSurplusNeeded (..)
+
+    -- ** Internals
+    , distributeSurplusDelta
+    )
+    where
```

Comment:
At first glance, this module seems to relate to two separate concepts:

- functions for computing the sizes of `Coin` values, orthogonal to any particular transaction balancing algorithm.
- an algorithm for distributing surplus value produced during the process of running the transaction balancing algorithm.

Perhaps these two concepts could be separated out into two separate modules?

- ` Internal.Cardano.Write.Tx.Balance.Size.Coin`
    - functions for computing the sizes of `Coin` values:
        - `sizeOfCoin`
        - `costOfIncreasingCoin`
        - `maximumCostOfIncreasingCoin`

- ` Internal.Cardano.Write.Tx.Balance.Surplus`
    - an algorithm for distributing surplus value:
        - `distributeSurplus`
        - `distributeSurplusDelta`
        - `ErrMoreSurplusNeeded`

Reply:
> Balance.Surplus

I thought to myself "that actually works; we _are_ distributing _balance_" before misremembering your suggestion as `Balance.Distribute` and creating #4542 😅 🤔 I think I do find `Distribute` to be the more descriptive term though so I'd lean towards it. What do you think?

Reply:
> Perhaps these two concepts could be separated out into two separate modules?

They are intertwined and I thought bundling them both under the `Size.Coin` module name made decent enough sense. Without variable-length coin encoding we wouldn't need `distributeSurplus`. However, we can start by calling it something else as per above, not to be blocked on the `Size` hiearchy idea.